### PR TITLE
CRM-20264 make pan_truncation save on contribution form, additional payment, event

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
@monishdeb here is what it takes to get the event form working - see the last commit

Note I moved
    $this->assignPaymentRelatedVariables();

to the shared preProcess function rather than just add it to one form

---

 * [CRM-20264: Store CC type and last 4 digits from Contribution form](https://issues.civicrm.org/jira/browse/CRM-20264)

### CODE REVIEW

File: CRM/Contribute/Form/AbstractEditPayment.php [hunk 1]
```diff
@@ -221,6 +221,7 @@ public function preProcess() {
     CRM_Core_Resources::singleton()->addVars('coreForm', array('contact_id' => (int) $this->_contactID));
     $this->_action = CRM_Utils_Request::retrieve('action', 'String', $this, FALSE, 'add');
     $this->_mode = empty($this->_mode) ? CRM_Utils_Request::retrieve('mode', 'String', $this) : $this->_mode;
+    $this->assignPaymentRelatedVariables();
```
Comment (eileenmcnaughton) [inline CRM/Contribute/Form/AbstractEditPayment.php]: moved here from being on many forms

File: CRM/Contribute/Form/AbstractEditPayment.php [hunk 2]
```diff
@@ -645,4 +646,17 @@ protected function getBillingDefaults($defaults) {
     return array_merge($defaults, $billingDefaults);
   }
 
+  /**
+   * Get the default payment instrument id.
+   *
+   * @return int
+   */
+  protected function getDefaultPaymentInstrumentId() {
```
Comment (eileenmcnaughton) [inline CRM/Contribute/Form/AbstractEditPayment.php]: moved here from Contribution form

File: CRM/Contribute/Form/AdditionalPayment.php [hunk 3]
```diff
@@ -77,10 +82,10 @@ public function preProcess() {
 
     parent::preProcess();
     $this->_id = CRM_Utils_Request::retrieve('id', 'Positive', $this, TRUE);
+    // @todo don't set this - rely on parent $this->contactID
     $this->_contactId = CRM_Utils_Request::retrieve('cid', 'Positive', $this, TRUE);
     $this->_component = CRM_Utils_Request::retrieve('component', 'String', $this, TRUE);
     $this->_view = CRM_Utils_Request::retrieve('view', 'String', $this, FALSE);
-    $this->_action = CRM_Utils_Request::retrieve('action', 'String', $this, FALSE);
```
Comment (eileenmcnaughton) [inline CRM/Contribute/Form/AdditionalPayment.php]: set in preProcess

File: CRM/Contribute/Form/AdditionalPayment.php [hunk 4]
```diff
@@ -137,8 +142,6 @@ public function preProcess() {
 
     list($this->_contributorDisplayName, $this->_contributorEmail) = CRM_Contact_BAO_Contact_Location::getEmailDetails($this->_contactId);
 
-    $this->assignPaymentRelatedVariables();
```
Comment (eileenmcnaughton) [inline CRM/Contribute/Form/AdditionalPayment.php]: moved to parent preProcess

File: CRM/Contribute/Form/AdditionalPayment.php [hunk 5]
```diff
@@ -330,6 +333,8 @@ public function postProcess() {
    *
    */
   public function submit($submittedValues) {
+    $this->_params = $submittedValues;
```
Comment (eileenmcnaughton) [inline CRM/Contribute/Form/AdditionalPayment.php]: these fixes are part of the test pr - that PR could be merged separately as it stands alone & is from Pradeeps other pr

File: CRM/Contribute/Form/AdditionalPayment.php [hunk 6]
```diff
@@ -462,11 +462,6 @@ public function processCreditCard($submittedValues) {
     $this->_params["state_province-{$this->_bltID}"] = $this->_params["billing_state_province-{$this->_bltID}"] = CRM_Core_PseudoConstant::stateProvinceAbbreviation($this->_params["billing_state_province_id-{$this->_bltID}"]);
     $this->_params["country-{$this->_bltID}"] = $this->_params["billing_country-{$this->_bltID}"] = CRM_Core_PseudoConstant::countryIsoCode($this->_params["billing_country_id-{$this->_bltID}"]);
 
-    if ($this->_paymentProcessor['payment_type'] & CRM_Core_Payment::PAYMENT_TYPE_CREDIT_CARD) {
```
Comment (eileenmcnaughton) [inline CRM/Contribute/Form/AdditionalPayment.php]: all done in beginPostProcess

File: CRM/Contribute/Form/AdditionalPayment.php [hunk 7]
```diff
@@ -495,14 +490,6 @@ public function processCreditCard($submittedValues) {
       $this->_bltID
     ));
 
-    $date = CRM_Utils_Date::format($params['credit_card_exp_date']);
```
Comment (eileenmcnaughton) [inline CRM/Contribute/Form/AdditionalPayment.php]: done in beginPostProcess

File: CRM/Contribute/Form/AdditionalPayment.php [hunk 8]
```diff
@@ -665,6 +651,11 @@ public function emailReceipt(&$params) {
    */
   public function testSubmit($params, $creditCardMode = NULL, $enitityType = 'contribute') {
     $this->_bltID = 5;
+    // Required because processCreditCard calls set method on this.
```
Comment (eileenmcnaughton) [inline CRM/Contribute/Form/AdditionalPayment.php]: just the test harness in this function

File: CRM/Contribute/Form/Contribution.php [hunk 9]
```diff
@@ -253,7 +253,6 @@ public function preProcess() {
     $this->assign('showCheckNumber', TRUE);
 
     $this->_fromEmails = CRM_Core_BAO_Email::getFromEmail();
-    $this->assignPaymentRelatedVariables();
```
Comment (eileenmcnaughton) [inline CRM/Contribute/Form/Contribution.php]: moved to parent::preProcess

File: CRM/Event/Form/Participant.php [hunk 10]
```diff
@@ -204,20 +208,19 @@ public function getDefaultEntity() {
...
-    $this->_contactId = CRM_Utils_Request::retrieve('cid', 'Positive', $this);
-    $this->_mode = CRM_Utils_Request::retrieve('mode', 'String', $this);
+    // @todo eliminate this duplication.
+    $this->_contactId = $this->_contactID;
     $this->_eID = CRM_Utils_Request::retrieve('eid', 'Positive', $this);
     $this->_context = CRM_Utils_Request::retrieve('context', 'String', $this);
     $this->assign('context', $this->_context);
 
-    if ($this->_contactId) {
-      $displayName = CRM_Contact_BAO_Contact::displayName($this->_contactId);
```
Comment (eileenmcnaughton) [inline CRM/Event/Form/Participant.php]: done in parent::preProcess

File: CRM/Event/Form/Participant.php [hunk 11]
```diff
@@ -204,20 +208,19 @@ public function getDefaultEntity() {
    * @return void
    */
   public function preProcess() {
+    parent::preProcess();
     $this->_showFeeBlock = CRM_Utils_Array::value('eventId', $_GET);
     $this->assign('showFeeBlock', FALSE);
     $this->assign('feeBlockPaid', FALSE);
 
-    $this->_contactId = CRM_Utils_Request::retrieve('cid', 'Positive', $this);
-    $this->_mode = CRM_Utils_Request::retrieve('mode', 'String', $this);
```
Comment (eileenmcnaughton) [inline CRM/Event/Form/Participant.php]: done in parent::preProcess

File: CRM/Event/Form/Participant.php [hunk 12]
```diff
@@ -1340,20 +1342,22 @@ public function postProcess() {
...
           'contribution_status_id',
-          'receive_date',
           'check_number',
           'campaign_id',
+          'pan_truncation',
         );
 
         foreach ($recordContribution as $f) {
-          $contributionParams[$f] = CRM_Utils_Array::value($f, $params);
+          $contributionParams[$f] = CRM_Utils_Array::value($f, $this->_params);
```
Comment (eileenmcnaughton) [inline CRM/Event/Form/Participant.php]: use of $this->_params includes the copied variables & reflects the fact the parent form & other classes use this. There is a tonne of confusion about the various arrays :-( I think we are moving back to $this->_params... at the moment

File: CRM/Member/Form.php [hunk 13]
```diff
@@ -303,8 +302,6 @@ protected function setContextVariables($params) {
       }
     }
 
-    $this->assignPaymentRelatedVariables();
```
Comment (eileenmcnaughton) [inline CRM/Member/Form.php]: in parent preProcess

**Comment by monishdeb** [comment]
Reviewed the changes. Good to merge

**Comment by mattwire** [comment]
This removes constant CRM_Core_Payment::PAYMENT_TYPE_DIRECT_DEBIT but it's still used in CRM/Contribute/Form/ContributionBase.php on line 577

**Comment by colemanw** [comment]
@eileenmcnaughton did you mean to remove that constant?

**Comment by colemanw** [comment]
Nevermind I see @monishdeb added it back.

**Comment by JoeMurray** [comment]
@mattwire thanks for catching that, but it's been fixed

**Pull request merged**